monkey/common/event_queue/event_queue.py

```diff
@@ -0,0 +1,53 @@
+from typing import Any, Callable, Sequence
+
+from pubsub import pub
+
+from common.events import AbstractEvent
+
+from .i_event_queue import IEventQueue
+
+
+class EventQueue(IEventQueue):
+    @staticmethod
+    def subscribe_all(subscriber: Callable[..., Any]):
+        pub.subscribe(listener=subscriber, topicName=pub.ALL_TOPICS)
+
+    @staticmethod
+    def subscribe_types(types: Sequence[AbstractEvent], subscriber: Callable[..., Any]):
+        for event_type in types:
+            # pypubsub.pub.subscribe needs a string as the topic/event name
+            event_type_name = event_type.__name__
+            pub.subscribe(listener=subscriber, topicName=event_type_name)
+
+    @staticmethod
+    def subscribe_tags(tags: Sequence[str], subscriber: Callable[..., Any]):
+        for tag in tags:
+            pub.subscribe(listener=subscriber, topicName=tag)
+
+    @staticmethod
+    def publish(event: AbstractEvent, data: Any = None):
+        data = data if data else {}
+
+        # publish to event type's topic
+        event_type_name = event.__name__
```

Comment:
I'm not sure this is a good idea. This couples the Event data structure to event implementation. For example, why does the event type needs to conform to pythons class naming standards. It's also less readable, when I go to the `AbstractEvent` I don't see the event type and I need to figure out that event type is actually the class name. Lastly the serialization. It's not enough to get the data, you also need to somehow include the class name and omit it after the serialization.

Reply:
> why does the event type needs to conform to pythons class naming standards

It doesn't need to, but pypubsub requires that the topic be a string. Since we can't pass the class as a topic, using the class's name seems appropriate. The alternative is to define an "event_type" field in every event class, just like we're defining tags. I prefer just using the name.


> It's not enough to get the data, you also need to somehow include the class name and omit it on de-serialization.

Why do you need the class's name? The subscriber that is subscribed to an event type (a class) knows what it's subscribed to and what to do with the data that it gets when the event is triggered. I don't see why it would need that information again.

Reply:
> why does the event type needs to conform to pythons class naming standards

It doesn't need to, but pypubsub requires that the topic be a string. Since we can't pass the class as a topic, using the class's name seems appropriate. The alternative is to define an "event_type" field in every event class, just like we're defining tags. I prefer just using the name.


> It's not enough to get the data, you also need to somehow include the class name and omit it on de-serialization.

Why do you need the class's name? The subscriber that is subscribed to an event type (a class) knows what it's subscribed to and what to do with the data that it gets when the event is triggered. I don't see why it would need that information again.

The only time we'll need the class name (for now) is when sending the event data to the API endpoint. In that case, we can do something like this with the subscriber when working on #2177:
```python3
>>> from pubsub import pub
>>> def listener(arg1, topic=pub.AUTO_TOPIC):
...     print(arg1, topic.getName(), sep="\n")
...
>>> pub.subscribe(listener, "some_topic")
(<pubsub.core.listener.Listener object at 0x0000012831600A88>, True)
>>> pub.sendMessage("some_topic", **{"arg1": 123})
123
some_topic
```

Reply:
> Why do you need the class's name?

Because we're talking about json serialization and json does not serialize classes, it serializes dicts. And currently, event type is not part of that dict. I think there are two ways to go on this: 
1. Mainly use only one event class. Add type and content parameters. This allows easy serialization/deserialization and dinamic content
2. We define new event for each type. Event content structure is based on the event type and needs to be strictly defined with a schema. This allows easier operations with event objects.

If we go with nr2, then class name as type is better, otherwise dedicated type param is better.  nr2 is harder to extend though, because you'll need to define schemas for new events and won't be able to just send generic event to the API

Reply:
For now, we'll leave it as-is. When we're finished implementing stolen credentials, we'll have an example that cuts through the entire stack. At that time, we can reevaluate. In other words, consider this a tracer bullet.